packages/@aws-cdk/aws-codepipeline/lib/private/rich-action.ts

```diff
@@ -0,0 +1,36 @@
+import { Aws, ResourceEnvironment } from '@aws-cdk/core';
+import { IAction } from '../action';
+
+/**
+ * Helper routines to work with Actions
+ *
+ * Can't put these on Action themselves since we only have an interface
+ * and every library would need to reimplement everything (there is no
+ * `ActionBase`).
+ *
+ * So here go the members that should have gone onto the Action class
+ * but can't.
+ *
+ * It was probably my own idea but I don't want it anymore:
+ * https://github.com/aws/aws-cdk/issues/10393
+ */
+export class RichAction {
+  constructor(private readonly action: IAction) {
+  }
+
+  /**
+   * The environment this action runs in
+   */
+  public get env(): ResourceEnvironment {
+    return {
+      account: (this.action.actionProperties.role?.env.account
+        ?? this.action.actionProperties?.resource?.env.account
+        ?? this.action.actionProperties.account
+        ?? Aws.ACCOUNT_ID),
```

Comment:
Hmmm... is this actually a sane default? If the account is not know, I feel like it should be `undefined` here, which means "this is the same as the Pipeline account" (doesn't matter if the Pipeline account is `AWS::AccountId`, or a specific account ID).

Reply:
It's a good point and I was in doubt about this initially, but I settled on this eventually.

The motivation here is that the cases aren't actually different, if you think about the end result: what ends up in the pipeline structure?

Even we defaulted to `undefined` here, the end result could still be `Aws.ACCOUNT_ID` (it could come from the Role, the Resource or the `account` property).  So now we have two cases to deal with: `undefined | Aws.ACCOUNT_ID`. 

*Both of those cases render the same*: they both end up rendering `undefined` into the template structure. 

So why would we have the calling side do case analysis to fold those cases into each other? It feels like you're retaining fidelity and retaining information (from this side of the implementation), but I would argue it's fake fidelity. It's more cases that the caller has to deal with, that he's *not actually* interested in (i.e., more chances for bugs to sneak in).

In fact, I just feel that in principle there's no value to, when you ask an Action: "what is your account", for it to say "I don't know".  Especially since the answer would be treated as "oh well then it's the same as my account" anyway. The only correct answer is "any account".

Reply:
In my opinion, there is an edge case here, and it is a Pipeline in a Stack with a specific account/region, but an Action with `AWS::AccountId` / `AWS::Region` set as its account/region. In my opinion, that is incorrect, and we should error out in that case.

I'm not saying this validation needs to be done in this PR, but if we know there is a case that calls for it, this API should be capable of expressing it.